TestFoundation/xdgTestHelper/main.swift

```diff
@@ -55,15 +60,65 @@ class XDGCheck {
     }
 }
 
-if let arg = ProcessInfo.processInfo.arguments.last {
-    if arg == "--xdgcheck" {
-        XDGCheck.run()
-    }
-    if arg == "--getcwd" {
-        print(FileManager.default.currentDirectoryPath)
+
+// Used by TestProcess.test_interrupt()
+func signalTest() {
+    signal(SIGINT, SIG_IGN)
+    signal(SIGCONT, SIG_IGN)
+    signal(SIGTERM, { signum in exit(99) })
+
+    var signals: [String] = []
+    let semaphore = DispatchSemaphore(value: 0)
+    let sQueue = DispatchQueue(label: "signal queue")
+
+    let sigintSource = DispatchSource.makeSignalSource(signal: SIGINT, queue: sQueue)
+    sigintSource.setEventHandler(handler: DispatchWorkItem(block: {
+        signals.append("SIGINT")
+        semaphore.signal()
+    }))
+    sigintSource.resume()
+
+    let sigcontSource = DispatchSource.makeSignalSource(signal: SIGCONT, queue: sQueue)
+    sigcontSource.setEventHandler(handler: DispatchWorkItem(block: {
+        signals.append("SIGCONT")
+        semaphore.signal()
+    }))
+    sigcontSource.resume()
+
+    func getSignal() -> String? {
+        return sQueue.sync {
+            return signals.isEmpty ? nil : signals.removeFirst()
+        }
     }
-    if arg == "--echo-PWD" {
-        print(ProcessInfo.processInfo.environment["PWD"] ?? "")
+
+    // On Linux, print() doesnt currently flush the output over the pipe so use
+    // write() for now. On macOS, print() works fine.
+    write(1, "Ready\n", 6)
+
+    // 3 second timeout
+    let endTime = DispatchTime.now().uptimeNanoseconds + 3_000_000_000
+    while true {
+        let timeout = DispatchTime(uptimeNanoseconds: endTime)
+        guard semaphore.wait(timeout: timeout) == .success else {
+            print("Timedout")
+            exit(127)
+        }
+        if let signal = getSignal() {
+            let msg = "Signal: \(signal)\n"
+            _ = msg.withCString {
+                write(1, $0, msg.count)
+            }
+        }
     }
 }
 
+
+if let arg = ProcessInfo.processInfo.arguments.last {
+    switch(arg) {
+    case "--xdgcheck":      XDGCheck.run()
+    case "--getcwd":        print(FileManager.default.currentDirectoryPath)
+    case "--echo-PWD":      print(ProcessInfo.processInfo.environment["PWD"] ?? "")
+    case "--signal-test":   signalTest()
+    default:                break
+    }
+}
```

Comment:
I think we're going to have to rename this helper tool at some point :)

(no need to do it now)

Reply:
Does this logic look correct: 
```
    open func suspend() -> Bool {
        guard isRunning && processIdentifier > 0 else {
            return false
        }

        if suspendCount == 0 {
            if kill(processIdentifier, SIGSTOP) == 0 {
                suspendCount += 1
            }
        } else {
            suspendCount += 1
        }
        return true
    }

    open func resume() -> Bool {
        guard isRunning && processIdentifier > 0 else {
            return true
        }

        if suspendCount == 1 {
            if kill(processIdentifier, SIGCONT) == 0 {
                suspendCount -= 1
            }
        } else {
            suspendCount -= 1
        }
        return true
    }
 ```

The tests ran OK on Linux/macOS and Darwin native so I think the counting is correct.